### PR TITLE
Rails5 support and query params bug fix

### PR DESCRIPTION
**Pull request by ismasan**
## What
- Make sure it works with Rails 5
- Refactor into smaller instance-level functionality while keeping original class-level API
## How

`'action_controller/metal/strong_parameters'` needs to be required explicitly for Rails 5 to work.

The class-level public API of `.path_template_for` and `.url_template_for` delegates to an instance of `Parser`.

The parser uses an instance of `Resolver`, which wraps all interaction with Rails APIs. This is to make it easier to support new Rails versions in the future and decouple Rails from the rest of the parsing code.

The resolver's public API is

```
#has_route?(router_name String) Boolean
#generate(helper_method String, helper_name String, params Hash) String
```

`Resolver.instance` returns an instance that works with Rails 4 or 5. In the future new resolver classes can be implemented for compatibility with new Rails versions or maybe other frameworks.

### CODE REVIEW

**Comment by ismasan** [comment]
not sure why the build's failing on Travis. I tried it locally with different Ruby versions and everything's green.

**Comment by ismasan** [comment]
This PR also fixes a bug where, when you supply _all_ query parameters, it was including a trailing `{&}` at the end of the string.

### Before

```ruby
Zooplankton.path_template_for(:root, %w(a b), "a" => "foo", "b" => "bar") # /?a=foo&b=bar{&}
```

### After

```ruby
Zooplankton.path_template_for(:root, %w(a b), "a" => "foo", "b" => "bar") # /?a=foo&b=bar
```

**Comment by benhamill** [comment]
This is pretty awesome. Thanks!

Do you mind adding a test for the new bug fix?

I'm trying to think about hoe to handle the version thing. Like... should I just release a new version that locks to Rails `~> 5.0`? Or cover both major versions in one (for now)?

Maybe could you also edit the Travis file to test against the two Rails versions? I'll have a look and see if I can tell why it's failing.

**Comment by benhamill** [comment]
Ah.

```
An error occurred while installing rack (2.0.1), and Bundler cannot continue.
```

So that's fun. Might be we should update the Ruby versions in the Travis file, too. Whatchya think?

**Comment by benhamill** [comment]
Oh. And would you mind adding a new `## Unreleased` section to `CHANGES.md` and give yourself credit however you'd like to be credited?

**Comment by ismasan** [comment]
Hi.

There's a test case for the bug fix already https://github.com/benhamill/zooplankton/pull/7/commits/f469126954bda3e4ceceeb7da06a93352358b275

Re. versions there aren't any API changes and it works with both Rails 4 and 5 so maybe a minor version upgrade is enough? I'll look into using both Rails versions in Travis. Maybe pinning to specific versions will fix the Rack issue. (?)

**Comment by ismasan** [comment]
Phew, that took some time!

So Travis is working correctly now. I've set up the test matrix to run [Ruby 2.0.0 / Rails 4] and [Ruby 2.2.2 / Rails 5]. The problem was that Rails5's Rack version needs Ruby >= 2.2.2.

I've added and documented the [Appraisal](https://github.com/thoughtbot/appraisal) gem in dev mode so we can run tests agains both versions of Rails in development.

I've also documented the changes under `## Unreleased`.

**Comment by ismasan** [comment]
@benhamill let me know if you have any comments about this.

File: zooplankton.gemspec [hunk 1]
```diff
@@ -23,5 +23,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency "pry"
   spec.add_development_dependency "rspec", "~> 2.14"
 
-  spec.add_dependency "rails", "~> 4.0"
+  spec.add_dependency "rails", ">= 4.0"
```
Comment (benhamill) [inline zooplankton.gemspec]: I think we should break this out to be greater than or equal to `4.0`, but less than `6.0`:
```
spec.add_dependency "rails", [">= 4.0", "< 6.0"]
```
Right? Like a super twiddlewakka.

**Comment by benhamill** [comment]
This is super awesome. Thanks for the time you've put into this.

**Comment by benhamill** [comment]
@ismasan: Do you want push/commit access to this repo and publish rights for the gem? You have literally done more work on this library in the last week than I have done in about 2 years.

File: zooplankton.gemspec (hunk 1 above)
Comment (ismasan) [inline zooplankton.gemspec]: Makes sense. Will do.

Comment (ismasan) [inline zooplankton.gemspec]: Done.

**Comment by ismasan** [comment]
Thanks for the offer. I'm happy with the pull-request-to-your-repository workflow and I can't promise to help maintain / curate PRs long term (I'm doing a bad enough job at that on my open OS libraries already :(), but I'll be happy to help in as much as a can.

**Comment by benhamill** [comment]
Cool. This looks good to me. Thanks for all your work. I'm going to merge this to master and try to find some time to actually cut a new version soon. :sparkling_heart:

**Comment by ismasan** [comment]
Hi! Any chance we could push a new release with the latest changes? If you want you can give me publish rights and I'll do it.

**Comment by benhamill** [comment]
Ack. I haven't touched my personal dev machine in, well, months, obviously. I'm installing stuff now to do this. If that fails because I haven't been maintaining that machine's dev environment, I'll see about getting it to work from my work laptop. Sorry that "soon" ended up being a lie.

**Comment by benhamill** [comment]
@ismasan OK. Zooplankton 1.3.0 should be shipped. Thanks for the reminder.

**Comment by ismasan** [comment]
No worries, I know how it is :) Thanks a lot for this.

**Pull request merged**